src/index/am.rs

```diff
@@ -292,27 +297,90 @@ pub unsafe extern "C" fn amrescan(
     _n_orderbys: std::os::raw::c_int,
 ) {
     unsafe {
-        if (*scan).numberOfKeys != 0 {
-            pgrx::error!("vector search with attributes is not supported");
-        }
-        if (*scan).numberOfOrderBys == 0 {
-            pgrx::error!("vector search without a ORDER BY clause is not supported");
+        if (*scan).numberOfOrderBys < 0 || (*scan).numberOfOrderBys > 1 {
+            pgrx::error!(
+                "vector search with {} ORDER BY clauses is not supported",
+                (*scan).numberOfOrderBys
+            );
         }
-        if (*scan).numberOfOrderBys != 1 {
-            pgrx::error!("vector search with too many ORDER BY clauses is not supported");
+        if (*scan).numberOfKeys < 0 {
+            pgrx::error!(
+                "vector search with {} WHERE clauses is not supported",
+                (*scan).numberOfKeys
+            );
         }
         if !keys.is_null() && (*scan).numberOfKeys > 0 {
             std::ptr::copy(keys, (*scan).keyData, (*scan).numberOfKeys as _);
         }
         if !orderbys.is_null() && (*scan).numberOfOrderBys > 0 {
             std::ptr::copy(orderbys, (*scan).orderByData, (*scan).numberOfOrderBys as _);
         }
-        let main = (*scan).orderByData.add(0);
-        let value = (*main).sk_argument;
-        let is_null = ((*main).sk_flags & pgrx::pg_sys::SK_ISNULL as i32) != 0;
-        let vector = from_datum_to_vector(value, is_null);
+        (*scan).xs_recheck = false;
+        let orderby_vector = match (*scan).numberOfOrderBys {
+            0 => None,
+            1 => {
+                let data = (*scan).orderByData.add(0);
+                let value = (*data).sk_argument;
+                let is_null = ((*data).sk_flags & pgrx::pg_sys::SK_ISNULL as i32) != 0;
+                from_datum_to_vector(value, is_null)
+            }
+            _ => unreachable!(),
+        };
+        let (range_vector, threshold) = match (*scan).numberOfKeys {
+            0 => (None, None),
+            1 => {
+                let data = (*scan).keyData.add(0);
+                let value = (*data).sk_argument;
+                let is_null = ((*data).sk_flags & pgrx::pg_sys::SK_ISNULL as i32) != 0;
+                let (options, _) = am_options::options((*scan).indexRelation);
+                let (v, threshold) = from_datum_to_range(value, &options.vector, is_null);
+                if orderby_vector.is_none() || (orderby_vector.is_some() && v == orderby_vector) {
+                    (v, threshold)
+                } else {
+                    (*scan).xs_recheck = true;
+                    (None, None)
+                }
+            }
+            n if orderby_vector.is_some() => {
```

Comment:
```suggestion
            n => {
```

Since it should be checked foremost, there is no need to check it and there should not be the next match arm.